### PR TITLE
[VL] Add memory pool to track shuffle cached memory usage

### PR DESCRIPTION
**Pull request by marin-ma**
After this patch, VeloxShuffleWriter has three memory pool.
- `options_.memory_pool`
- `partitionBufferPool_` :  Used to track memory usage of split buffers. The actual allocation is delegated to `options_.memory_pool`.
- `payloadPool_` : Used to track memory allocation of Arrow IPC payloads. The actual allocation is delegated to `options_.memory_pool`.

We used maintain `options_.ipc_memory_pool`, which can be assigned as `LargeMemoryPool` if needed. `LargeMemoryPool` was introduced to alleviate the occurrence of page fault, and it was only used for allocating Arrow IPC payload. But since the creating IPC payload part (compressing buffers, etc) has been refactored many times, we haven't done performance testing around the latest implementation, so the `LargeMemoryPool` will be removed as well.

depends on #2982

### CODE REVIEW

File: shims/common/src/main/scala/io/glutenproject/GlutenConfig.scala [hunk 1]
```diff
@@ -774,6 +776,12 @@ object GlutenConfig {
       .booleanConf
       .createWithDefault(true)
 
+  val COLUMNAR_SHUFFLE_REALLOC_THRESHOLD =
+    buildConf("spark.gluten.sql.columnar.shuffle.realloc.threshold")
```
Comment (Yohahaha) [inline shims/common/src/main/scala/io/glutenproject/GlutenConfig.scala:780]: nit: would you like copy docs from Configuration.md?

Comment (marin-ma) [reply shims/common/src/main/scala/io/glutenproject/GlutenConfig.scala:780]: @Yohahaha This PR depends on #2982 Please check this commit only https://github.com/oap-project/gluten/pull/3159/commits/feb200942a9f505bbf7394a4a1117040fa3748f4

File: cpp/core/shuffle/ShuffleWriter.h [hunk 2]
```diff
@@ -288,17 +257,17 @@ class ShuffleWriter {
       : numPartitions_(numPartitions),
         partitionWriterCreator_(std::move(partitionWriterCreator)),
         options_(std::move(options)),
-        pool_(std::make_shared<ShuffleMemoryPool>(options_.memory_pool)),
+        splitBufferPool_(std::make_shared<ShuffleMemoryPool>(options_.memory_pool)),
```
Comment (zuochunwei) [inline cpp/core/shuffle/ShuffleWriter.h:260]: changing the type of `splitBufferPool_` to `ShuffleMemoryPool` may be better.

the lifecycle of `splitBufferPool_` is binding with ShuffleWriter.

File: cpp/velox/shuffle/VeloxShuffleWriter.cc [hunk 3]
```diff
@@ -343,6 +337,13 @@ std::shared_ptr<arrow::RecordBatch> makeUncompressedRecordBatch(
   }
   return arrow::RecordBatch::Make(writeSchema, 1, {arrays});
 }
+
+inline bool shrinkBeforeSpill(SplitState state) {
+  return state == SplitState::kStop;
+}
```
Comment (zuochunwei) [inline cpp/velox/shuffle/VeloxShuffleWriter.cc:343]: add a space line

File: cpp/velox/shuffle/VeloxShuffleWriter.cc [hunk 4]
```diff
@@ -343,6 +337,13 @@ std::shared_ptr<arrow::RecordBatch> makeUncompressedRecordBatch(
   }
   return arrow::RecordBatch::Make(writeSchema, 1, {arrays});
 }
+
+inline bool shrinkBeforeSpill(SplitState state) {
```
Comment (zuochunwei) [inline cpp/velox/shuffle/VeloxShuffleWriter.cc:341]: the `inline` is unnecessary, the compiler will make the decision when she sees the function defined in the inner of the class.

**Comment by zuochunwei** [comment]
Could you provide some benchmark data for this PR?
I am curious about how much heap memory is saved after this change and the performance changes.

**Comment by FelixYBW** [comment]
@zhztheplayer Let's add the buffers on OOM message like
shuffle write:
  spill buffer size:
  split buffer size:
  arrow::default memory pool size:

**Comment by zhztheplayer** [comment]
> @zhztheplayer Let's add the buffers on OOM message once this PR is merged. something like shuffle write: spill buffer size: split buffer size: arrow::default memory pool size:

We can just use different Arrow pools for different purposes? Do you think it's technically possible? @marin-ma

**Review by FelixYBW — APPROVED** [review]

**Pull request merged**